### PR TITLE
syncs: add ShardedMap.Mutate

### PR DESCRIPTION
**Pull request by bradfitz**
To let callers do atomic/CAS-like operations.

### CODE REVIEW

File: syncs/shardedmap.go [hunk 1]
```diff
@@ -59,9 +59,36 @@ func (m *ShardedMap[K, V]) Get(key K) (value V) {
 	return
 }
 
+// Mutate atomically mutates m[k] by calling mutator.
+//
+// The mutator function is called with the old value (or its zero value) and
+// whether it existed in the map and it returns the new value and whether it
+// should be added vs deleted from the map.
```
Comment (creachadair) [inline syncs/shardedmap.go:66]: maybe: added/kept or added/retained ?

File: syncs/shardedmap_test.go [hunk 2]
```diff
@@ -41,4 +41,41 @@ func TestShardedMap(t *testing.T) {
 	if g, w := m.Len(), 0; g != w {
 		t.Errorf("got Len %v; want %v", g, w)
 	}
+
+	// Mutation adding an entry.
+	if v := m.Mutate(1, func(was string, ok bool) (string, bool) {
```
Comment (creachadair) [inline syncs/shardedmap_test.go:46]: Is it worthwhile also verifying that the implementation made the desired change in the map?

**Review by creachadair — APPROVED** [review]

File: syncs/shardedmap.go (hunk 1 above)
Comment (bradfitz) [reply syncs/shardedmap.go:66]: clarified

File: syncs/shardedmap_test.go (hunk 2 above)
Comment (creachadair) [reply syncs/shardedmap_test.go:46]: Ah yeah, so it does.

**Review by dsnet — APPROVED** [review]
I like it.

**Pull request merged**